tmol/chemical/restypes.py

```diff
@@ -178,6 +178,54 @@ def _setup_path_distance(self):
         path_distance[path_distance == numpy.inf] = MAX_SIG_BOND_SEPARATION
         return path_distance
 
+    atom_paths_from_conn: numpy.ndarray = attr.ib()
+
+    @atom_paths_from_conn.default
+    def _setup_atom_paths_from_conn(self):
+        if len(self.connections) == 0:
+            return
+
+        n_conns = len(self.connections)
+        MAX_PATHS = 13  # TODO: if we keep this constant, it should be somewhere globally accessible...
+
+        atom_paths = numpy.full((n_conns, MAX_PATHS, 3), -1, dtype=numpy.int32)
+
+        def get_paths_length_3(connection):
+            paths = numpy.full((MAX_PATHS, 3), -1, dtype=numpy.int32)
+            # create a convenient datastructure for following connections
+            bondmap = {-1: []}
+            for bond in self.bond_indices:
+                if bond[0] not in bondmap.keys():
+                    bondmap[bond[0]] = []
+                bondmap[bond[0]].append(bond[1])
+
+            atom0 = self.atom_to_idx[connection.atom]
+            paths[0] = (atom0, -1, -1)
+
+            idx = 1
+            for atom1 in bondmap[atom0] + [-1] * (3 - len(bondmap[atom0])):
+                if atom1 != -1:
+                    # atoms.append((atom0, atom1, -1))
+                    paths[idx] = (atom0, atom1, -1)
+                idx += 1
+
+            for atom1 in bondmap[atom0] + [-1] * (3 - len(bondmap[atom0])):
```

Comment:
I'm not really sure I understand.

There is the potential for a full row of -1s between valid paths, but I think this is necessary if we want to avoid having to iterate through tensor to find the locations of specific paths down in the C++. 

As you say, the point of the system is to keep the 1-atom paths, 2-atom paths, 3-atom paths in deterministic locations. 

If there is a simpler way of doing this, I can make modifications.

Reply:
I've added a comment to try and make what is happening here more clear.